### PR TITLE
[FLINK-28778][SQL/API] Bulk fetch of table and column statistics for given partitions

### PR DESCRIPTION
**Pull request by JingGe**
## What is the purpose of the change

Currently the statistics information about tables can only be fetched from the catalog by each given partition iteratively. Since getting statistics information from catalogs is a very heavy operation, in order to improve the query performance, we’d better provide functionality to fetch the statistics information of a table for all given partitions in one shot.

Based on the manual performance test, for 2000 partitions, the cost will be improved from 10s to 2s. The improvement result is 500%.

## Brief change log

- Implementations for new added methods will be done in all classes that implement the Catalog interface. 
- For the concrete HiveCatalog, __HIVE_DEFAULT_PARTITION__ will be taken care of while fetching the column statistics. - - All currently supported types for partition column will be taken into consideration.
- The logic of getting TableStates in FlinkRecomputeStatisticsProgram will be optimized from calling the catalog iteratively to bulk fetch.
- Logic of TableStates conversion will be consolidated into the CatalogTableStatisticsConverter class to make the domain design a little bit cleaner. 

## Verifying this change

This change is already covered by existing tests, such as HiveCatalogHiveMetadataTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by JingGe** [comment]
@flinkbot run azure

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java [hunk 1]
```diff
@@ -720,6 +720,25 @@ public CatalogTableStatistics getPartitionStatistics(
         return result != null ? result.copy() : CatalogTableStatistics.UNKNOWN;
     }
 
+    /**
+     * For in memory catalog, just iteratively call {@link #getPartitionStatistics(ObjectPath,
+     * CatalogPartitionSpec)} for each partition.
+     */
+    @Override
+    public List<CatalogTableStatistics> bulkGetPartitionStatistics(
```
Comment (godfreyhe) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:728]: the implementation is totally same the the default one, I think we can remove it

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/CatalogTableStatisticsConverter.java [hunk 2]
```diff
@@ -61,6 +63,27 @@ public static TableStats convertToTableStats(
...
+                        .map(p -> CatalogTableStatisticsConverter.convertToTableStats(p, null))
+                        .reduce((s1, s2) -> s1.merge(s2));
+
+        final Optional<TableStats> columnStatsMergedTableStats =
+                catalogColumnStatisticsList.stream()
+                        .map(
+                                p ->
+                                        CatalogTableStatisticsConverter.convertToTableStats(
+                                                CatalogTableStatistics.EMPTY, p))
+                        .reduce((s1, s2) -> s1.merge(s2));
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/CatalogTableStatisticsConverter.java:80]: We can just convert a CatalogTableStatistics and its corresponding  CatalogColumnStatistics into a TableStats, and then call TableStats#merge. If that, CatalogTableStatistics.EMPTY, StatisticDataUtils can be removed. There's no calculation here, and even CatalogColumnStatisticDataBuilder is no needed. This pr will become very simple

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java [hunk 3]
```diff
@@ -438,6 +440,38 @@ public CatalogTableStatistics getPartitionStatistics(
         return CatalogTableStatistics.UNKNOWN;
     }
 
+    @Override
+    public List<CatalogTableStatistics> bulkGetPartitionStatistics(
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:444]: I think we can just use the default implementation, and remove this method.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkRecomputeStatisticsProgram.java [hunk 4]
```diff
@@ -183,36 +181,29 @@ private TableStats getPartitionsTableStats(
...
-            CatalogColumnStatistics partitionColStat =
-                    catalog.getPartitionColumnStatistics(tablePath, spec);
-            TableStats stats =
-                    CatalogTableStatisticsConverter.convertToTableStats(
-                            partitionStat, partitionColStat);
-            return Optional.of(stats);
+
+            final List<CatalogPartitionSpec> partitionSpecs =
+                    partition.stream()
+                            .map(p -> new CatalogPartitionSpec(p))
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkRecomputeStatisticsProgram.java:200]: nit: .map(CatalogPartitionSpec::new)

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 5]
```diff
@@ -1699,6 +1716,69 @@ public CatalogTableStatistics getPartitionStatistics(
...
+            ObjectPath tablePath, List<CatalogPartitionSpec> partitionSpecs)
+            throws PartitionNotExistException, CatalogException {
+
+        return getTableAndPartitions(tablePath, partitionSpecs).f1.stream()
+                .map(p -> createCatalogTableStatistics(p.getParameters()))
+                .collect(Collectors.toList());
+    }
+
+    /** @return Partitions and names of partitions. */
+    private Tuple3<Table, List<Partition>, List<String>> getTableAndPartitions(
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:1730]: We can just split this method into two methods, one is to get the HiveTable, and another is to get partitions. That could make the result type more simple and readable

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 6]
```diff
@@ -1734,6 +1814,152 @@ public CatalogColumnStatistics getPartitionColumnStatistics(
...
+        final CatalogColumnStatisticsDataBinary.Builder binaryBuilder =
+                CatalogColumnStatisticsDataBinary.builder();
+
+        for (String partitionName : partitionNames) {
+            String[] partName = partitionName.split(Path.SEPARATOR);
+            String[] partVal = partName[partName.length - 1].split("=");
+            String value = partVal[1];
+            if (value.equals(HIVE_DEFAULT_PARTITION)) {
+                try {
+                    nullCount =
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:1908]: Not set nullCount

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java [hunk 7]
```diff
@@ -240,6 +246,84 @@ void testAlterPartitionColumnStatistics() throws Exception {
...
+                tableColumnStatistics.get(tableColumnStatistics.size() - 1);
+        CatalogColumnStatisticsDataLong columnStatisticsDataLong =
+                (CatalogColumnStatisticsDataLong)
+                        catalogColumnStatistics.getColumnStatisticsData().get(partitionName);
+
+        assertThat(columnStatisticsDataLong.getNdv()).isEqualTo(catalogPartitionSpecList.size());
+        assertThat(columnStatisticsDataLong.getMin()).isEqualTo(0);
+        assertThat(columnStatisticsDataLong.getMax()).isEqualTo(partitionFileSize - 1);
+
+        final List<CatalogPartitionSpec> catalogPartitionSpecs = catalog.listPartitions(path1);
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java:319]: This line can be removed ?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java [hunk 8]
```diff
@@ -240,6 +246,84 @@ void testAlterPartitionColumnStatistics() throws Exception {
...
+
+        final CatalogColumnStatistics catalogColumnStatistics =
+                tableColumnStatistics.get(tableColumnStatistics.size() - 1);
+        CatalogColumnStatisticsDataLong columnStatisticsDataLong =
+                (CatalogColumnStatisticsDataLong)
+                        catalogColumnStatistics.getColumnStatisticsData().get(partitionName);
+
+        assertThat(columnStatisticsDataLong.getNdv()).isEqualTo(catalogPartitionSpecList.size());
+        assertThat(columnStatisticsDataLong.getMin()).isEqualTo(0);
+        assertThat(columnStatisticsDataLong.getMax()).isEqualTo(partitionFileSize - 1);
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java:317]: other columns' statistics does not been verified

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java [hunk 9]
```diff
@@ -311,11 +314,7 @@ void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ignoreIfE
     void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean ignoreIfNotExists)
             throws TableNotExistException, CatalogException;
 
-    /**
-     * If true, tables which do not specify a connector will be translated to managed tables.
-     *
-     * @see CatalogBaseTable.TableKind#MANAGED
-     */
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java:318]: Unnecessary change ?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java [hunk 10]
```diff
@@ -566,6 +565,31 @@ CatalogTableStatistics getPartitionStatistics(
...
+     *
+     * @param tablePath path of the table
+     * @param partitionSpecs partition specs of partitions that will be used to filter out all other
+     *     unrelated statistics, i.e. the statistics fetch will be limited within the given
+     *     partitions
+     * @return list of statistics of given partitions
+     * @throws PartitionNotExistException if one partition does not exist
+     * @throws CatalogException in case of any runtime exception
+     */
+    default List<CatalogTableStatistics> bulkGetPartitionStatistics(
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java:579]: This method should be tested via GenericInMemoryCatalog

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @JingGe , I left some comments

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 11]
```diff
@@ -1734,6 +1814,152 @@ public CatalogColumnStatistics getPartitionColumnStatistics(
...
+                                    .getRowCount();
+                } catch (Exception e) {
+                    LOG.info(
+                            "Fail to get getPartitionColumnStatistics for partition {}={}",
+                            partitionedByColumn.getName(),
+                            HIVE_DEFAULT_PARTITION);
+                }
+            } else {
+                final String type = partitionedByColumn.getType();
+                if ("int".equalsIgnoreCase(type)) {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:1924]: Avoid to use "int" directly. May be we can do the check using LogicalType like `HiveStatsUtil#getColumnStatisticsData`

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java (hunk 8 above)
Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java:317]: +1

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/stats/CatalogColumnStatisticsDataBoolean.java [hunk 12]
```diff
@@ -54,4 +54,45 @@ public CatalogColumnStatisticsDataBoolean copy() {
...
+            this.nullCount = nullCount;
+            return this;
+        }
+
+        public Builder properties(Map<String, String> properties) {
+            this.properties = properties;
+            return this;
+        }
+
+        public Builder increasTrueCount(Long trueCount) {
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/stats/CatalogColumnStatisticsDataBoolean.java:82]: nit:
```suggestion
        public Builder increaseTrueCount(Long trueCount) {
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/stats/StatisticDataUtils.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog.stats;
+
+/** Utils class for catalog column statistics. */
+public class StatisticDataUtils {
+
+    public static Double acculmulateAverage(
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/stats/StatisticDataUtils.java:24]: nit:
```suggestion
    public static Double accumulateAverage(
```

**Review by luoyuxia — COMMENTED** [review]
@JingGe Thanks for contribution. I left some comments.

**Comment by wuchong** [comment]
Continue with #20501
